@here/harp-mapview/lib/AnimatedExtrusionHandler.ts

```diff
@@ -33,341 +38,220 @@ export class AnimatedExtrusionHandler {
+    private getTreeRootMortonCode(tile: Tile): number {
+        const rootLevel = tile.dataSource.getDataZoomLevel(this.minZoomLevel);
+        return tile.tileKey.changedLevelTo(rootLevel).mortonCode();
     }
 
-    private stopExtrusionAnimation(): void {
-        this.m_mapView.removeEventListener(MapViewEventNames.AfterRender, this.animateExtrusion);
+    private isTileTreeAdded(tile: Tile): boolean {
+        return this.m_tileTreeSizeMap.has(this.getTreeRootMortonCode(tile));
     }
 
-    private animateExtrusion = (event?: RenderEvent) => {
-        if (this.m_animatedExtrusionState !== AnimatedExtrusionState.Playing) {
-            if (this.m_animatedExtrusionState === AnimatedExtrusionState.Started) {
-                this.m_animatedExtrusionState = AnimatedExtrusionState.Playing;
-            } else {
-                return;
-            }
+    private animateExtrusion() {
+        if (this.m_state !== AnimatedExtrusionState.Started) {
+            return;
         }
 
         const currentTime = Date.now();
-        if (
-            this.m_animatedExtrusionStartTime === undefined ||
-            this.m_animatedExtrusionStartTime <= 0
-        ) {
-            this.m_animatedExtrusionStartTime = currentTime;
+        if (this.m_startTime < 0) {
+            this.m_startTime = currentTime;
         }
 
-        const timeProgress = Math.min(
-            currentTime - this.m_animatedExtrusionStartTime,
-            this.m_animatedExtrusionDuration
-        );
+        const duration = this.duration;
+        const timeProgress = Math.min(currentTime - this.m_startTime, duration);
 
-        this.extrusionRatio = MathUtils.easeInOutCubic(
+        const extrusionRatio = MathUtils.easeInOutCubic(
             ExtrusionFeatureDefs.DEFAULT_RATIO_MIN,
             ExtrusionFeatureDefs.DEFAULT_RATIO_MAX,
-            timeProgress / this.m_animatedExtrusionDuration
+            timeProgress / duration
         );
+        this.setExtrusionRatio(extrusionRatio);
 
-        if (timeProgress >= this.m_animatedExtrusionDuration) {
-            this.m_animatedExtrusionState = AnimatedExtrusionState.Finished;
-            this.stopExtrusionAnimation();
+        if (timeProgress >= duration) {
+            this.m_state = AnimatedExtrusionState.Finished;
         }
 
-        this.m_tile.dataSource.requestUpdate();
-    };
+        this.m_mapView.update();
+    }
+
+    private setExtrusionRatio(value: number, force: boolean = false) {
+        this.m_tileMap.forEach(materials => {
+            this.setTileExtrusionRatio(materials, value, force);
+        });
+    }
+
+    private setTileExtrusionRatio(
+        materials: ExtrusionFeature[],
+        value: number,
+        force: boolean = false
+    ) {
+        materials.forEach(material => {
+            if ((material.extrusionRatio ?? 0) < 1 || force) {
+                material.extrusionRatio = value;
```

Comment:
this only sets the extrusionRatio if less than 1, however an extrusionRatio of 1 is also valid?

Reply:
that condition is to avoid animating again tiles that are fully extruded (extrusionRatio === 1).

Reply:
ok, thanks.